Check if Task(Future) is canceled.

fujitatomoya opened the pull request:
closes https://github.com/ros2/rclpy/issues/1099

fujitatomoya commented on rclpy/test/test_executor.py:
```diff
@@ -273,6 +273,27 @@ async def coroutine():
         self.assertTrue(future.done())
         self.assertEqual('Sentinel Result', future.result())
 
+    def test_create_task_coroutine_cancel(self) -> None:
```
with this change, all the other tests are passing but this additional one.

this test generates the following warning, eventually this test case fails... i spent some time to see what is going on here, but still not sure about this. requesting help for this.

`colcon test --event-handlers console_direct+ --packages-select rclpy --ctest-args -R test_executor`

<details><summary>PluggyTeardownRaisedWarning</summary>

```bash
13: ../../src/ros2/rclpy/rclpy/test/test_executor.py .......F............... [ 82%]
13: .....                                                                    [100%]
13:
13: =================================== FAILURES ===================================
13: ________________ TestExecutor.test_create_task_normal_function _________________
13:
13:     @pytest.hookimpl(hookwrapper=True, tryfirst=True)
13:     def pytest_runtest_call() -> Generator[None, None, None]:
13: >       yield from unraisable_exception_runtest_hook()
13:
13: /usr/lib/python3/dist-packages/_pytest/unraisableexception.py:88:
13: _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
13:
13:     def unraisable_exception_runtest_hook() -> Generator[None, None, None]:
13:         with catch_unraisable_exception() as cm:
13:             yield
13:             if cm.unraisable:
13:                 if cm.unraisable.err_msg is not None:
13:                     err_msg = cm.unraisable.err_msg
13:                 else:
13:                     err_msg = "Exception ignored in"
13:                 msg = f"{err_msg}: {cm.unraisable.object!r}\n\n"
13:                 msg += "".join(
13:                     traceback.format_exception(
13:                         cm.unraisable.exc_type,
13:                         cm.unraisable.exc_value,
13:                         cm.unraisable.exc_traceback,
13:                     )
13:                 )
13: >               warnings.warn(pytest.PytestUnraisableExceptionWarning(msg))
13: E               pytest.PytestUnraisableExceptionWarning: Exception ignored in: <coroutine object TestExecutor.test_create_task_coroutine_cancel.<locals>.coroutine at 0x7f198fffa980>
13: E
13: E               Traceback (most recent call last):
13: E                 File "/usr/lib/python3.12/warnings.py", line 553, in _warn_unawaited_coroutine
13: E                   warn(msg, category=RuntimeWarning, stacklevel=2, source=coro)
13: E               RuntimeWarning: coroutine 'TestExecutor.test_create_task_coroutine_cancel.<locals>.coroutine' was never awaited
13:
13: /usr/lib/python3/dist-packages/_pytest/unraisableexception.py:78: PytestUnraisableExceptionWarning
13:
13: During handling of the above exception, another exception occurred:
13:
13: cls = <class '_pytest.runner.CallInfo'>
13: func = <function call_runtest_hook.<locals>.<lambda> at 0x7f198e051f80>
13: when = 'call'
13: reraise = (<class '_pytest.outcomes.Exit'>, <class 'KeyboardInterrupt'>)
13:
13:     @classmethod
13:     def from_call(
13:         cls,
13:         func: "Callable[[], TResult]",
13:         when: "Literal['collect', 'setup', 'call', 'teardown']",
13:         reraise: Optional[
13:             Union[Type[BaseException], Tuple[Type[BaseException], ...]]
13:         ] = None,
13:     ) -> "CallInfo[TResult]":
13:         """Call func, wrapping the result in a CallInfo.
13:
13:         :param func:
13:             The function to call. Called without arguments.
13:         :param when:
13:             The phase in which the function is called.
13:         :param reraise:
13:             Exception or exceptions that shall propagate if raised by the
13:             function, instead of being wrapped in the CallInfo.
13:         """
13:         excinfo = None
13:         start = timing.time()
13:         precise_start = timing.perf_counter()
13:         try:
13: >           result: Optional[TResult] = func()
13:
13: /usr/lib/python3/dist-packages/_pytest/runner.py:341:
13: _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
13: /usr/lib/python3/dist-packages/_pytest/runner.py:262: in <lambda>
13:     lambda: ihook(item=item, **kwds), when=when, reraise=reraise
13: /usr/lib/python3/dist-packages/pluggy/_hooks.py:501: in __call__
13:     return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
13: /usr/lib/python3/dist-packages/pluggy/_manager.py:119: in _hookexec
13:     return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
13: _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
13:
13: hook_name = 'pytest_runtest_call'
13: hook_impl = <HookImpl plugin_name='unraisableexception', plugin=<module '_pytest.unraisableexception' from '/usr/lib/python3/dist-packages/_pytest/unraisableexception.py'>>
13: e = PytestUnraisableExceptionWarning('Exception ignored in: <coroutine object TestExecutor.test_create_task_coroutine_canc...\nRuntimeWarning: coroutine \'TestExecutor.test_create_task_coroutine_cancel.<locals>.coroutine\' was never awaited\n')
13:
13:     def _warn_teardown_exception(
13:         hook_name: str, hook_impl: HookImpl, e: BaseException
13:     ) -> None:
13:         msg = "A plugin raised an exception during an old-style hookwrapper teardown.\n"
13:         msg += f"Plugin: {hook_impl.plugin_name}, Hook: {hook_name}\n"
13:         msg += f"{type(e).__name__}: {e}\n"
13:         msg += "For more information see https://pluggy.readthedocs.io/en/stable/api_reference.html#pluggy.PluggyTeardownRaisedWarning"  # noqa: E501
13: >       warnings.warn(PluggyTeardownRaisedWarning(msg), stacklevel=5)
13: E       pluggy.PluggyTeardownRaisedWarning: A plugin raised an exception during an old-style hookwrapper teardown.
13: E       Plugin: unraisableexception, Hook: pytest_runtest_call
13: E       PytestUnraisableExceptionWarning: Exception ignored in: <coroutine object TestExecutor.test_create_task_coroutine_cancel.<locals>.coroutine at 0x7f198fffa980>
13: E
13: E       Traceback (most recent call last):
13: E         File "/usr/lib/python3.12/warnings.py", line 553, in _warn_unawaited_coroutine
13: E           warn(msg, category=RuntimeWarning, stacklevel=2, source=coro)
13: E       RuntimeWarning: coroutine 'TestExecutor.test_create_task_coroutine_cancel.<locals>.coroutine' was never awaited
13: E
13: E       For more information see https://pluggy.readthedocs.io/en/stable/api_reference.html#pluggy.PluggyTeardownRaisedWarning
13:
13: /usr/lib/python3/dist-packages/pluggy/_callers.py:49: PluggyTeardownRaisedWarning
13: - generated xml file: /root/ros2_ws/colcon_ws/build/rclpy/test_results/rclpy/test_executor.xunit.xml -
13: =========================== short test summary info ============================
13: FAILED ../../src/ros2/rclpy/rclpy/test/test_executor.py::TestExecutor::test_create_task_normal_function
13: ========================= 1 failed, 27 passed in 7.21s =========================
13: Exception ignored in: <function Executor.__del__ at 0x7f19977f7240>
13: Traceback (most recent call last):
13:   File "/root/ros2_ws/colcon_ws/src/ros2/rclpy/rclpy/rclpy/executors.py", line 262, in __del__
13:     self._sigint_gc.destroy()
13:   File "/root/ros2_ws/colcon_ws/src/ros2/rclpy/rclpy/rclpy/signals.py", line 70, in destroy
13:     with self.handle:
13: test_rclpy._rclpy_pybind11.InvalidHandle: cannot use Destroyable because destruction was requested
13: -- run_test.py: return code 1
13: -- run_test.py: verify result file '/root/ros2_ws/colcon_ws/build/rclpy/test_results/rclpy/test_executor.xunit.xml'
1/1 Test #13: test_executor ....................***Failed    7.82 sec
```

</summary>

nadavelkabets replied:
Seems like the test raised `RuntimeWarning: coroutine 'TestExecutor.test_create_task_coroutine_cancel.<locals>.coroutine' was never awaited`.

Previously, because of the bug, a cancelled task was not removed from the executor and was never garbage collected.
After your change, `future.cancel()` causes the executor to drop the task and the coroutine is never yielded. Some time later, the coroutine object is garbage collected, and since the coroutine is not closed, the exception is raised.

The error "coroutine was never awaited" is common and can be easily replicated:
```python
async def foo():
	pass

coro = foo()
coro.__del__()

RuntimeWarning: coroutine 'foo' was never awaited
```
Calling `coro.close()` solves this issue by throwing `GeneratorExit` into the coroutine https://docs.python.org/3/reference/datamodel.html#coroutine.close

There are two possible solutions I see:
1. Copying asyncio and cancelling the task by throwing a CancelledError exception into the coroutine. 
The main benefit of this approach is the possibility of user code inside the coroutine to catch the exception and perform cleanup.
https://github.com/python/cpython/blob/061e50f196373d920c3eaa3718b9d0553914e006/Lib/asyncio/tasks.py#L270-L273
https://github.com/python/cpython/blob/061e50f196373d920c3eaa3718b9d0553914e006/Lib/asyncio/tasks.py#L283-L307

```python
def cancel(self):
	try:
		if not self._done and iscoroutine(self._handle):
			self._handle.throw(CancelledError())
	except CancelledError as exc:
		self._exception = exc
	finally:
		super().cancel()

```
2. Calling `coro.close()`
```python
def cancel(self):
	if not self._done and iscoroutine(self._handle):
		  self._handle.close()
	super().cancel()

```

Also, I see that you called `asyncio.sleep(1)` in your test. It doesn't really matter in this case since this portion of the code is never executed, but `await asyncio.sleep(1)` is not possible in rclpy, since the implementation calls for `asyncio.get_running_loop()` but no asyncio loop is initialized or running (`asyncio.sleep(0)` is a little different as it just yields).
https://github.com/python/cpython/blob/061e50f196373d920c3eaa3718b9d0553914e006/Lib/asyncio/tasks.py#L688-L705

haudren-woven replied:
I wholeheartedly agree with the first suggestion above :point_up: If we are implementing cancellation, I think we should follow asyncio's lead, as it would also help with #1098 . However, if we want to go the whole way, we'd also need to introduce a way to `await` subscriptions and timers (with e.g. an async iterator).

weber-niklas replied:
I agree with @haudren-woven on this. Creating usage close to asyncio's implementation would also create a better entry path for new ROS users/ developers. I personally very much enjoyed the CancelledError Exception approach when working with asyncio.

fujitatomoya replied:
https://github.com/ros2/rclpy/pull/1394 implements option 2 which is to close the coroutine via Task.cancel. and the follow up discussion is required on https://github.com/ros2/rclpy/pull/1392

nadavelkabets commented on rclpy/rclpy/task.py:
```diff
@@ -61,7 +61,7 @@ def __del__(self) -> None:
     def __await__(self) -> Generator[None, None, Optional[T]]:
         # Yield if the task is not finished
-        while not self._done:
+        while not self._done and not self._cancelled:
```
Asyncio tasks work a little differently, they have 3 possible states and are considered done if they are not pending:
https://github.com/python/cpython/blob/ee0746af7d7cfc6cc25441726034e4fea4bcf7e5/Lib/asyncio/base_futures.py#L7-L10
https://github.com/python/cpython/blob/403410fa1be036214efa7955127911e5592910db/Lib/asyncio/futures.py#L177-L183
@fujitatomoya what do you think?

fujitatomoya replied:
this is addressed by https://github.com/ros2/rclpy/pull/1394

nadavelkabets commented:
Please backport this bugfix into humble/jazzy.
Cancelling tasks is a crucial feature of any event loop, and for a project of mine I had to come up with complex and incomplete solutions to make that work.

weber-niklas commented:
Hey,
any updates here?
I think @fujitatomoya was pretty close to make it work.
Maybe if we go with @nadavelkabets suggestion number 2:

```python

def cancel(self):
	if not self._done and iscoroutine(self._handle):
		  self._handle.close()
	super().cancel()
```

We can at least fix the bug at hand and concentrate on improving on the functionality after?
I am currently having a pretty hard time using Tasks since a bunch of it is broken. Would love to get a fix out which makes the feature usable. Thanks for the work you guys put in <3

fujitatomoya commented:
@nadavelkabets @weber-niklas sorry guys for being late to get back to this, i will try to allocate some time for this in next week.

nadavelkabets commented:
> @nadavelkabets @weber-niklas sorry guys for being late to get back to this, i will try to allocate some time for this in next week.

I created PR #1392 to fix the unawaited coroutine issue based on my first suggestion. It's not yet complete since we need to figure out how to handle the CancelledException in a way that won't crash the executor.

I believe we should first go with a solution that we can backport into jazzy.
Do you think suggestion 1 is viable for a backport, or should we start with suggestion 2 and add suggestion 1 for the next version?

fujitatomoya commented on rclpy/rclpy/task.py:
```diff
@@ -78,15 +84,15 @@ def cancelled(self) -> bool:
         :return: True if the task is finished or raised while it was executing
         """
-        return self._done
+        return self._state != FutureState.PENDING
```
@nadavelkabets 

To keep the consistent behavior including backports for jazzy and humble, probably we want to keep this as following?

```suggestion
        return self._state == FutureState.FINISHED
```

fujitatomoya commented:
@nadavelkabets again thank you very much for helping this one moving forward 👍 

i have already https://github.com/ros2/rclpy/pull/1394 merged in this PR.

> I believe we should first go with a solution that we can backport into jazzy.

agree, this is also broken in jazzy and humble.

> Do you think suggestion 1 is viable for a backport, or should we start with suggestion 2 and add suggestion 1 for the next version?

IMO, we can go with suggestion 2 (coroutine close, that is currently implemented in this PR) to avoid surprising the application with new exception generated by task cancel. that said it does not break the user space, user application can have the fix and also no user code change is required.

And then probably we can move on to https://github.com/ros2/rclpy/pull/1392 for rolling branch. i think having the capability for clean-up in the user callback makes sense for canceling tasks.

> It's not yet complete since we need to figure out how to handle the CancelledException in a way that won't crash the executor.

i will look into this for sure and leave some comments.

nadavelkabets commented on rclpy/rclpy/executors.py:
```diff
@@ -610,6 +620,8 @@ def _wait_for_ready_callbacks(
                 with self._tasks_lock:
                     # Get rid of any tasks that are done
                     self._tasks = list(filter(lambda t_e_n: not t_e_n[0].done(), self._tasks))
+                    # Get rid of any tasks that are cancelled
+                    self._tasks = list(filter(lambda t_e_n: not t_e_n[0].cancelled(), self._tasks))
```
Missed this line in my PR, not needed anymore since done() now returns true if cancelled.
There are some other duplicate checks for done and cancelled, I'll give the code a second pass to look for them.

nadavelkabets commented:
@fujitatomoya please review #1396, it includes some fixes I forgot to add in my last PR.
Is there anything I can do to help expedite this PR so we can review and merge it in the near future?

fujitatomoya commented:
@nadavelkabets thank for the PR against this one, can you also take a look at the unresolved comment here? that is directly related to https://github.com/ros2/rclpy/pull/1396.

nadavelkabets commented:
I created a new PR #1397 that should close your comments @fujitatomoya.

fujitatomoya commented on rclpy/rclpy/task.py:
```diff
@@ -114,12 +125,17 @@ def set_result(self, result: T) -> None:
         :param result: The output of a long running task.
         """
         with self._lock:
+            if not self._pending():
+                raise RuntimeError(f"Cannot set result from state: {self._state}")
+            
```
this breaks the following test (possibly userspace):

https://github.com/ros2/rclpy/blob/296425c4183135c048dff8fd22a4c0d44b39b0f4/rclpy/test/test_rosout_subscription.py#L114-L115

fujitatomoya replied:
fixed in https://github.com/ros2/rclpy/pull/1377/commits/6cd294b906ef3604235a9a9153fdb6e06262cdfa

fujitatomoya commented:
@nadavelkabets can you double-check https://github.com/ros2/rclpy/pull/1377/commits/6cd294b906ef3604235a9a9153fdb6e06262cdfa just in case? that just removes some flake8 errors, and removed new exception that you introduced with previous commit. do you think this is okay?

nadavelkabets commented on rclpy/rclpy/task.py:
```diff
@@ -78,15 +89,15 @@ def cancelled(self) -> bool:
         """
-        Indicate if the task has finished executing.
+        Indicate if the task has finished or cancelled executing.
 
-        :return: True if the task is finished or raised while it was executing
+        :return: True if the task is finished, cancelled or raised while it was executing
```
This is no longer true, we should revert the changes

fujitatomoya replied:
ah gosh, i knew that and i forgot that... 😓 fixed in https://github.com/ros2/rclpy/pull/1377/commits/628a9ee02cb3bf7ebaafb1bc48d73b104e630e0a

nadavelkabets commented:
> @nadavelkabets can you double-check [6cd294b](https://github.com/ros2/rclpy/commit/6cd294b906ef3604235a9a9153fdb6e06262cdfa) just in case? that just removes some flake8 errors, and removed new exception that you introduced with previous commit. do you think this is okay?

Looks good to me.
I think this PR is ready to move forward.

fujitatomoya commented:
Pulls: ros2/rclpy#1377
Gist: https://gist.githubusercontent.com/fujitatomoya/640f3a1c4e6f8072674dbe10b4778e7a/raw/b5a624de3bc2d1b1582cc1d000d22c2618a58681/ros2.repos
BUILD args:  --packages-above-and-dependencies rclpy
TEST args:  --packages-above rclpy
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/15053
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=22369)](http://ci.ros2.org/job/ci_linux/22369/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=16632)](http://ci.ros2.org/job/ci_linux-aarch64/16632/)
* Linux-rhel [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=1901)](http://ci.ros2.org/job/ci_linux-rhel/1901/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=23161)](http://ci.ros2.org/job/ci_windows/23161/)

tfoote commented on rclpy/rclpy/executors.py:
```diff
@@ -321,14 +321,24 @@ def spin_until_future_complete(
+                self._context.ok() and
+                not future.done() and
+                not future.cancelled()
+                and not self._is_shutdown
```
small nit, can you keep the `and` location consistent. I think the new style recommendation is after the line break for easier readability.

tfoote commented on rclpy/rclpy/executors.py:
```diff
@@ -321,14 +321,24 @@ def spin_until_future_complete(
-            while self._context.ok() and not future.done() and not self._is_shutdown:
+            while (
+                self._context.ok() and
+                not future.done() and
```
Suggesting logical operator at the beginning here too.

fujitatomoya commented:
@tfoote thanks for the review, i will address them.

besides that, CI RHEL and Windows are not happy with `ImportError: cannot import name 'StrEnum' from 'enum'`

nadavelkabets commented on rclpy/rclpy/task.py:
```diff
@@ -31,14 +32,19 @@ def _fake_weakref() -> None:
     return None
 
 
+class FutureState(StrEnum):
+    """States defining the lifecycle of a future."""
+
+    PENDING = auto()
+    CANCELLED = auto()
+    FINISHED = auto()
```
Suggested replacement for StrEnum:
```python
class FutureState(str, Enum):
    """States defining the lifecycle of a future."""

    PENDING = "PENDING"
    CANCELLED = "CANCELLED"
    FINISHED = "FINISHED"
```

nadavelkabets commented on rclpy/rclpy/task.py:
```diff
@@ -12,6 +12,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+from enum import auto, StrEnum
```
```suggestion
from enum import Enum
```

nadavelkabets commented:
> besides that, CI RHEL and Windows are not happy with `ImportError: cannot import name 'StrEnum' from 'enum'`

Good catch. StrEnum was only added in python3.11. 
I added suggestions for a fix.
@fujitatomoya should I open a new PR?

fujitatomoya commented:
@nadavelkabets no that is fine, i can patch them today.

fujitatomoya commented:
@tfoote @nadavelkabets addressed review comments, can you check https://github.com/ros2/rclpy/pull/1377/commits/fc3dd67f3f4d1bebfce14f3edd32c79bfc71c2cf?

fujitatomoya commented:
Pulls: ros2/rclpy#1377
Gist: https://gist.githubusercontent.com/fujitatomoya/bed30f370697213781f69ecce74a03e0/raw/b5a624de3bc2d1b1582cc1d000d22c2618a58681/ros2.repos
BUILD args:  --packages-above-and-dependencies rclpy
TEST args:  --packages-above rclpy
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/15055
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=22371)](http://ci.ros2.org/job/ci_linux/22371/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=16634)](http://ci.ros2.org/job/ci_linux-aarch64/16634/)
* Linux-rhel [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=1903)](http://ci.ros2.org/job/ci_linux-rhel/1903/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=23163)](http://ci.ros2.org/job/ci_windows/23163/)

tfoote commented on rclpy/test/test_task.py:
```diff
@@ -322,6 +322,34 @@ def cb(fut):
+    def test_set_result_on_done_future_without_exception(self) -> None:
+        f = Future()
+        f.set_result(None)
+        self.assertTrue(f.done())
```
To get coverage we should check the negatives too. 
```suggestion
        self.assertTrue(f.done())
        self.assertFalse(f.cancelled())
```

tfoote commented on rclpy/test/test_task.py:
```diff
@@ -322,6 +322,34 @@ def cb(fut):
+        f.set_result(None)
+        self.assertTrue(f.done())
+        f.set_result(None)
+        self.assertTrue(f.done())
```
```suggestion
        self.assertTrue(f.done())
        self.assertFalse(f.cancelled())
```

tfoote commented on rclpy/test/test_task.py:
```diff
@@ -322,6 +322,34 @@ def cb(fut):
+    def test_set_result_on_cancelled_future_without_exception(self) -> None:
+        f = Future()
+        f.cancel()
+        self.assertTrue(f.cancelled())
```
```suggestion
        self.assertTrue(f.cancelled())
        self.assertFalse(f.done())
```

tfoote commented on rclpy/test/test_task.py:
```diff
@@ -322,6 +322,34 @@ def cb(fut):
+
+    def test_set_exception_on_done_future_without_exception(self) -> None:
+        f = Future()
+        f.set_result(None)
```
Can we check the exception isn't there before and is after like this?
```suggestion
        f.set_result(None)
        self.assertIsNone(f.exception())
```

tfoote commented on rclpy/test/test_task.py:
```diff
@@ -322,6 +322,34 @@ def cb(fut):
+    def test_set_exception_on_cancelled_future_without_exception(self) -> None:
+        f = Future()
+        f.cancel()
+        self.assertTrue(f.cancelled())
```
```suggestion
        self.assertTrue(f.cancelled())
        self.assertIsNone(f.exception())
```

fujitatomoya commented:
@nadavelkabets @weber-niklas @haudren-woven @tonygoldman 

just a quick heads up. thank you very much for being patient on this PR.

this PR is almost ready to merge in rolling. once i confirm the all CI goes green, i will merge this to rolling branch 1st. since the original issue created on released distro (actually Foxy), and this is really nice to fix to downstream distros Jazzy and Humble, i will try to backport this after a couple of weeks soak time in rolling.

and special thanks to @nadavelkabets for helping me to push forward this fix 👍

nadavelkabets commented:
Awesome! It’s great news for any ‘async ros’ user!
@fujitatomoya thank you for dedicating the time and effort to this.
Honestly I’ve been waiting for this fix for the last two years…
It’s never too late😃

fujitatomoya commented:
windows is unstable, but i do not think failures are related to this PR.

retry CI (windows only):
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=23164)](http://ci.ros2.org/job/ci_windows/23164/)

fujitatomoya commented:
windows CI is green, merging.

fujitatomoya commented:
@Mergifyio backport humble jazzy

fujitatomoya commented:
@nadavelkabets @weber-niklas @haudren-woven @tonygoldman

Just FYI, jazzy and humble backports are completed.

https://github.com/ros2/rclpy/pull/1403
https://github.com/ros2/rclpy/pull/1404
